core/backend/src/ECSchemaFromIModel.ts

```diff
@@ -0,0 +1,23 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+import { IModelError, IModelStatus } from "@itwin/core-common";
+import { IModelDb } from "./IModelDb";
+
+/**
+ * Helper to get schema json from an iModel.
+ * @param iModel An iModel to get schemas from
+ * @returns A function that gets a schema from the input iModel by schema name or undefined if the schema is not found.
+ * @throws [IModelError]($core-common) if the schema is found but fails to load
+ * @alpha
+ */
+export const getSchemaJsonFromIModel = (iModel: IModelDb) => {
+  return (schemaName: string) => {
+    const val = iModel.nativeDb.getSchema(schemaName);
```

Comment:
please convert the native function to return a schema object and get rid of this method in backend.

Reply:
I wrapped the call to nativeDb in a method so that we didn't have callers outside of the core packages calling an internal method.

The SchemaLoader works like:
```ts
const schemaLoader = new SchemaLoader(getSchemaJsonFromIModel(iModelDb));
```

Without this method the caller would have to do this:
```ts
const schemaLoader = new SchemaLoader(iModelDb.nativeDb.getSchema));
```

I will add schema interfaces so I can return the object instead of just the string.  The interfaces will have to be duplicated between ecschema-metadata and backend but they are relatively small so that should be OK.

Reply:
There is no problem with **internal** code in our packages outside of core-backend accessing `iModelDb.native`, since we re-export its type from core-backend. But it is @internal so only internal code can access it.

Reply:
fine, just export it with it's public signature from whatever package implements it. That shouldn't be core-backend. 

*This* method does absolutely nothing other than unpack the useless "return" structure. It has no reason to exist.

Reply:
see [here](https://bentleycs.visualstudio.com/iModelTechnologies/_git/imodel02/pullrequest/263015?_a=files&path=/iModelJsNodeAddon/IModelJsNative.cpp) where I did this for you (although you can leave out the "async" if that turns out to be unnecessary)

Reply:
@kabentley I retested the async code path in that PR and found a bug in the test suite that was impacting the numbers.  I created a new C++ PR because the merge was very messy and updated the itwinjs-core PR so they should both be ready to go.

[itwinjs-core PR](https://github.com/iTwin/itwinjs-core/pull/3851)
[c++ PR](https://bentleycs.visualstudio.com/iModelTechnologies/_git/imodel02/pullrequest/271724)

Reply:
@kabentley I tried to remove the function from core-backend entirely but the IModelSchemaLoader is used by iTwin.js based connectors.  So both our docs, our code outside of itwinjs-core and 3rd party code would have to reference the internal api.

Without the function they would have to do this:
```ts
const loader = new SchemaLoader((schemaName: string) => { return iModel.nativeDb.getSchemaProps(schemaName); } );
```

With the function they do:
```ts
const loader = new SchemaLoader(makeSchemaPropsGetterFromIModel(iModel));
```

Both options are more complex than the old loader code but with the function it is still reasonable.

Reply:
I'm sorry, I just don't understand why this is hard. What's wrong with making a class `IModelSchemaLoader` in ecchema-metadata that takes an `IModelDb` as an argument?

Reply:
ecschema-metadata is used on the frontend so it cannot have a dependency on the backend

Reply:
where is IModelSchemaLoader?

Reply:
There is no longer an IModelSchemaLoader.  I refactored it into a generic SchemaLoader that takes a function that gets SchemaProps for a given schema name.  This is nice because the SchemaLoader now works with any source of SchemaProps and reduces the need for someone to write a ISchemaLocater which is more complex.

Callers that want to load schemas from an iModel must combine the generic SchemaLoader with the function that I exposed in core-backend that returns SchemaProps from an iModel.

Reply:
it seems to me you've added a useless level of indirection. Why take a function-that-gets-schemaprops instead of schemaprops?

Reply:
It gets used like this:
```ts
        const loader = new SchemaLoader(makeSchemaPropsGetterFromIModel(iModel));
        const schema1 = schemaLoader.getSchema("TestSchema1");
        const schema2 = schemaLoader.getSchema("TestSchema2");
```

so you do not have the SchemaProps when you construct the SchemaLoader or the names of the schemas that you will need

Reply:
What would be wrong with:
```ts
const loader = new SchemaLoader();
const schema1 = loader.load(iModel.getSchemaProps("TestSchema1"));
const schema2 = loader.load(iModel.getSchemaProps("TestSchema2"));
```
in IModelDb

```ts
public getSchemaProps(name: string) { return this.nativeDb.getSchema(name);}
```
way easier to use and understand.

Reply:
schemas reference each other so loading one schema may require loading N other schemas.

Reply:
then document it like this:
```ts
        const loader = new SchemaLoader((name: sting) => iModel.getSchemaProps(name));
        const schema1 = schemaLoader.getSchema("TestSchema1");
        const schema2 = schemaLoader.getSchema("TestSchema2");
```
I strongly object to a whole source file to wrap one nice clean function with a function-that-calls it.

Reply:
ok, I can expose getSchemaProps as a public method on IModelDb

Reply:
Done, IModelDb now exposes a getSchemaProps method.  This does look nicer.